### ISSUE TITLE
Deserialization of the state machine context data may fail due to Kryo problem but SSM doesn't care

### ISSUE THREAD

**Issue by tjormola** [spring-projects/spring-statemachine#551]
Please see this comment on Apache Kryo issue tracker: https://github.com/EsotericSoftware/kryo/issues/428#issuecomment-382882845

**Comment by jvalkeal** [comment]
Thanks for those screenshots, helps to understand what happens in kryo. I'm just wondering what might be the easiest way to reproduce this in tests so that we could try some workarounds(as i.e. that other user tweaked references setting in kryo). It's bad if this can happen with any other kryo usage as same mechanism is also used to persist machine with DB's.

**Comment by jvalkeal** [comment]
But yeah, we also need to do something about this silent failure.

**Comment by tjormola** [comment]
Just a side note, I wonder if it would be feasible to engineer more pluggable/configurable serialization mechanism for the state context? SSM could also ship with a Jackson and/or Java serialization based implementations that could configured/enabled on runtime by configuration properties with the Kryo implementation being the default so no breakage of existing systems and data would happen. A bit like e.g. the binder configuration mechanism and different binder implementations available for Spring Cloud Stream.. Maybe even leverage the message conversion mechanism of Spring Integration for converting the state context object to a byte array and vice versa.

**Comment by tjormola** [comment]
And for this issue and my project using SSM, I just ignore the distributed stuff for now and try to get it in shape inside a single JVM. When that is done, I will re-try with Kryo/Zookeeper and see if the problem persists after the many changes to the state machine since filing this issue. If the problem still exists, I'd happy to come up with a test case. Just don't have the time right now, sorry..

**Comment by jvalkeal** [comment]
Yeah, it might be one option to fully abstract serialisation. Originally it was just easiest route to get something relatively reliable working with kryo. I think it's already possible to use something else than kryo, but configuration is probably quite awkward as we do quite a bit of stuff with kryo to get something working oob.
